src/libDirectoryService/DSBlockPreProcessing.cpp

```diff
@@ -113,15 +114,23 @@ void DirectoryService::ComposeDSBlock(
                       << ", new difficulty " << std::to_string(difficulty));
     }
 
+    if (UpgradeManager::GetInstance().HasNewSW())
```

Comment:
From what I have seen, `HasNewSW` and `DownloadSW` are blocking calls involving curl downloading. Do we have any timeout protection that deals with unideal network connections so that the node won't be stuck?

Reply:
Cause we need to wait m_mediator.m_curSWInfo be updated, to compose a new DS block using latest SWInfo, we cannot detach HasNewSW() and DownloadSW() here. Will implement a timeout protection inside curl-related function in follow-up PR. (https://github.com/Zilliqa/Issues/issues/185)